Add J lexer

unsigned-wrong-wrong-int opened the pull request:
This adds a lexer for [J](https://www.jsoftware.com).

![image](https://user-images.githubusercontent.com/59091082/93428399-52a0be00-f8fa-11ea-9e22-99a1b961d23b.png)

jneen commented:
This looks great, thanks for the contribution! I have just a few comments.

jneen commented on lib/rouge/lexers/j.rb:
```diff
@@ -0,0 +1,231 @@
+            lex_str m
+          elsif m[3]
+            token Punctuation, m[1]
+            delegate @q_expr, m[2]
```
I see what you're doing here - but I wonder if there's a way to accomplish this without `delegate` - as in, match just the first opening `'` (and maybe some lookahead), then push `:q_expr`, and trust it to find the closing delimiter and pop in the right place. Are there instances of `'` that don't trigger this state?

jneen replied:
I mention this because often for string literals and similar it's desirable to lex *within* them, both for syntax-embedding purposes and also to be able to highlight escapes and other special characters. I don't know J, so it's possible this isn't a concern either.

unsigned-wrong-wrong-int commented:
Thanks for the review! I removed `delegate` and it works fine :)

I also added some primitives to `@primitive_table`. (AFAIK there are no further missing primitives.)

jneen commented:
Awesome, looks much better! :+1: from me.

jneen reviewed (approved):
Looks good to me!

pyrmont commented:
@unsigned-wrong-wrong-int Thanks for the submission! This will go out as part of v3.24.0 of Rouge. That's scheduled for release tonight/tomorrow (depending on your time zone) 🎉
